### DISCUSSION TITLE
Eigen include dir?

### DISCUSSION THREAD

**Discussion by paulmelis** [MrNeRF/LichtFeld-Studio#17] [Q&A] [answered]
First up, thanks for creating this project!

While trying to compile this I ran into an issue with the path to the eigen library. It's present correctly in `external/eigen` in my checkout (due to the recursive clone), yet the source files reference it under an `eigen3` dir, as shown by the error below. I noticed the `eigen3` cmake target, pointing to the `eigen` directory, but I don't think that works for `#include`, so those would still need `#include <eigen/...>`. Could it be that on your own development environment you also have a system-wide Eigen available (under something like `/usr/include/eigen3`) that is actually being used instead of the local copy in `external`? 

```
snellius paulm@gcn34 10:20 ~/c/gaussian-splatting-cuda/build$ ninja -j1
[1/21] Building CUDA object CMakeFiles/gaussian_splatting_cuda.dir/src/main.cu.o
FAILED: CMakeFiles/gaussian_splatting_cuda.dir/src/main.cu.o 
/sw/arch/RHEL8/EB_production/2022/software/CUDA/11.7.0/bin/nvcc -forward-unknown-to-host-compiler -DUSE_C10D_GLOO -DUSE_C10D_NCCL -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -I/home/paulm/c/gaussian-splatting-cuda/includes -I/home/paulm/c/gaussian-splatting-cuda/cuda_rasterizer -I/home/paulm/c/gaussian-splatting-cuda/external/tinyply/source -I/sw/arch/RHEL8/EB_production/2022/software/Python/3.10.4-GCCcore-11.3.0/include/python3.10 -I/home/paulm/c/gaussian-splatting-cuda/external/simple-knn -I/home/paulm/c/gaussian-splatting-cuda/external/json/include -isystem=/home/paulm/c/gaussian-splatting-cuda/external/libtorch/include -isystem=/home/paulm/c/gaussian-splatting-cuda/external/libtorch/include/torch/csrc/api/include -isystem=/sw/arch/RHEL8/EB_production/2022/software/CUDA/11.7.0/include -isystem=/home/paulm/c/gaussian-splatting-cuda/external/eigen -isystem=/home/paulm/c/gaussian-splatting-cuda/external/glm -DONNX_NAMESPACE=onnx_c2 -gencode arch=compute_80,code=sm_80 -Xcudafe --diag_suppress=cc_clobber_ignored,--diag_suppress=integer_sign_change,--diag_suppress=useless_using_declaration,--diag_suppress=set_but_not_used,--diag_suppress=field_without_dll_interface,--diag_suppress=base_class_has_different_dll_interface,--diag_suppress=dll_interface_conflict_none_assumed,--diag_suppress=dll_interface_conflict_dllexport_assumed,--diag_suppress=implicit_return_from_non_void_function,--diag_suppress=unsigned_compare_with_zero,--diag_suppress=declared_but_not_referenced,--diag_suppress=bad_friend_decl --expt-relaxed-constexpr --expt-extended-lambda -O3 -DNDEBUG --generate-code=arch=compute_52,code=[compute_52,sm_52] -O3 -use_fast_math -Xcompiler -Ofast -D_GLIBCXX_USE_CXX11_ABI=1 -std=c++17 -MD -MT CMakeFiles/gaussian_splatting_cuda.dir/src/main.cu.o -MF CMakeFiles/gaussian_splatting_cuda.dir/src/main.cu.o.d -x cu -c /home/paulm/c/gaussian-splatting-cuda/src/main.cu -o CMakeFiles/gaussian_splatting_cuda.dir/src/main.cu.o
In file included from /home/paulm/c/gaussian-splatting-cuda/includes/render_utils.cuh:5,
                 from /home/paulm/c/gaussian-splatting-cuda/src/main.cu:5:
/home/paulm/c/gaussian-splatting-cuda/includes/camera.cuh:4:10: fatal error: eigen3/Eigen/Dense: No such file or directory
    4 | #include <eigen3/Eigen/Dense>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```

Edit: also, the failing build line above uses `-isystem=/home/paulm/c/gaussian-splatting-cuda/external/eigen`, which implies Eigen could never come from that location, as that contains an `Eigen` subdir with all the headers. I.e. the include path would need to be `<Eigen/...>` without the `eigen3/` prefix to work.

**Comment by MrNeRF** [comment]
Hi Paul,

Thank you for bringing this to my attention. 

This is the first time I've encountered this issue, and from those I know who have successfully built it, no one has mentioned this problem. The path `-isystem=/home/paulm/c/gaussian-splatting-cuda/external/eigen` looks correct to me. I've checked it against my build folder, and it appears nearly identical. However, you've raised a valid point; I do have an `eigen3` folder on my system. This might actually be the root of the problem.

Inside the `external` folder, there's a `CMakeLists.txt` that pulls in the external libraries:

```
add_library(eigen3 INTERFACE IMPORTED GLOBAL)
target_include_directories(eigen3 INTERFACE eigen)
```

You could be onto something. I just tried to change  `<eigen3/Eigen/Dense>` to `<Eigen/Dense>`, and it built successfully on my end. Could you please test it with the Eigen includes on your side? There aren't many to change. If it works for you and builds without any issues, I'd appreciate it if you could open a pull request. However, if you'd prefer, and upon your confirmation, I can handle the update. Whichever route you choose, please let me know the outcome.

Best
Janusch

**Reply by paulmelis** [reply, marked as answer]
Hi Janusch, I had indeed already tested with changing `<eigen3/Eigen/...>` to `<Eigen/...>` and that fixes the build issue for me.

**Reply by paulmelis** [reply]
Just created pull request #18

**Reply by MrNeRF** [reply]
Nice Paul, thanks a lot. I will take a look and then merge it.
